independent-projects/arc/tests/src/test/java/io/quarkus/arc/test/producer/generic/WildcardProducerHierarchyTest.java

```diff
@@ -0,0 +1,99 @@
+package io.quarkus.arc.test.producer.generic;
+
+import io.quarkus.arc.Arc;
+import io.quarkus.arc.test.ArcTestContainer;
+import org.junit.jupiter.api.Assertions;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.RegisterExtension;
+
+import java.util.Date;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+import javax.enterprise.context.ApplicationScoped;
+import javax.enterprise.inject.Produces;
+import javax.enterprise.util.TypeLiteral;
+import javax.inject.Singleton;
+
+public class WildcardProducerHierarchyTest {
+
+    @RegisterExtension
+    public ArcTestContainer container = new ArcTestContainer(Producer.class);
+
+    @Test
+    public void testWildcardProducers() {
+        Function<String, Date> fn = Arc.container().instance(new TypeLiteral<Function<String, Date>>() {}).get();
+        Assertions.assertNotNull(fn.apply(String.valueOf(System.currentTimeMillis())));
+    }
+
+    @Singleton
+    static class Producer {
+
+        @Produces
+        @ApplicationScoped
+        public AsyncBiFunctionService.WithInner<String, Long, Date> inner() {
+            return new InnerImpl();
+        }
+
+        @Produces
+        @Singleton
+        public LocalService<? extends AsyncBiFunctionService.WithInner<String, Long, Date>> localInner(AsyncBiFunctionService.WithInner<String, Long, Date> service) {
```

Comment:
`LocalService<? extends AsyncBiFunctionService.WithInner<String, Long, Date>>` is not a legal bean type and should be removed from the set of bean types of this producer method.

Reply:
Why is that?

Reply:
https://docs.jboss.org/cdi/spec/2.0/cdi-spec.html#legal_bean_types